src/poetry/core/packages/dependency_group.py

```diff
@@ -22,35 +25,42 @@ def __init__(
         self._mixed_dynamic = mixed_dynamic
         self._dependencies: list[Dependency] = []
         self._poetry_dependencies: list[Dependency] = []
+        self._included_dependency_groups: dict[NormalizedName, DependencyGroup] = {}
 
     @property
     def name(self) -> str:
         return self._name
 
     @property
     def dependencies(self) -> list[Dependency]:
+        included_group_dependencies = self._resolve_included_dependency_groups()
+
         if not self._dependencies:
             # legacy mode
-            return self._poetry_dependencies
+            return self._poetry_dependencies + included_group_dependencies
         if self._mixed_dynamic and self._poetry_dependencies:
             if all(dep.is_optional() for dep in self._dependencies):
                 return [
                     *self._dependencies,
                     *(d for d in self._poetry_dependencies if not d.is_optional()),
+                    *included_group_dependencies,
                 ]
             if all(not dep.is_optional() for dep in self._dependencies):
                 return [
                     *self._dependencies,
                     *(d for d in self._poetry_dependencies if d.is_optional()),
+                    *included_group_dependencies,
                 ]
-        return self._dependencies
+        return self._dependencies + included_group_dependencies
 
     @property
     def dependencies_for_locking(self) -> list[Dependency]:
+        included_group_dependencies = self._resolve_included_dependency_groups()
```

Comment:
It might make no difference at the moment, but I think this is wrong because `_resolve_included_dependency_groups` uses `dependencies` instead of `dependencies_for_locking`. I think we need two versions of `_resolve_included_dependency_groups` (or pass a parameter that tells which dependencies to use).

Reply:
Actually I guess I need more information about what's the difference between `dependencies` and `dependencies_for_locking` :sweat_smile:

Reply:
I guess you do not need to know the difference to consider it wrong: It feels clear to me that nesting groups should not violate the distinction between the two.

Nevertheless: `dependencies` are used for metadata (when building), `dependencies_for_locking` are used for locking. This distinction is probably not that relevant for groups because we only need `dependencies_for_locking` and do not use `dependencies` but that does not mean that we should not care about implementing it correctly.

Reply:
I've attempted to address this issue in https://github.com/finswimmer/core/pull/2 & added a test for this case.